crates/vm/src/system/memory/paged_vec.rs

```diff
@@ -44,36 +44,109 @@ impl<T: Default + Clone, const PAGE_SIZE: usize> PagedVec<T, PAGE_SIZE> {
             None
         }
     }
-
     #[inline(always)]
     pub fn range_vec(&self, range: Range<usize>) -> Vec<T> {
-        let mut result = Vec::with_capacity(range.len());
-        for page_idx in (range.start / PAGE_SIZE)..range.end.div_ceil(PAGE_SIZE) {
-            let in_page_start = range.start.saturating_sub(page_idx * PAGE_SIZE);
-            let in_page_end = (range.end - page_idx * PAGE_SIZE).min(PAGE_SIZE);
-            if let Some(page) = self.pages[page_idx].as_ref() {
-                result.extend(page[in_page_start..in_page_end].iter().cloned());
+        let len = range.end - range.start;
+        // Create a vector for uninitialized values.
+        let mut result: Vec<MaybeUninit<T>> = Vec::with_capacity(len);
+        // SAFETY: We set the length and then initialize every element.
+        unsafe {
+            result.set_len(len);
+            let dst = result.as_mut_ptr() as *mut T;
+            let start_page = range.start / PAGE_SIZE;
+            let end_page = (range.start + len - 1) / PAGE_SIZE;
+            if start_page == end_page {
+                let offset = range.start % PAGE_SIZE;
+                if let Some(page) = self.pages[start_page].as_ref() {
+                    let src = page.as_ptr().add(offset);
+                    std::ptr::copy_nonoverlapping(src, dst, len);
+                } else {
+                    for i in 0..len {
+                        std::ptr::write(dst.add(i), T::default());
+                    }
+                }
             } else {
-                result.extend(vec![T::default(); in_page_end - in_page_start]);
+                debug_assert_eq!(start_page + 1, end_page);
+                let offset = range.start % PAGE_SIZE;
+                let first_part = PAGE_SIZE - offset;
+                if let Some(page) = self.pages[start_page].as_ref() {
+                    let src = page.as_ptr().add(offset);
+                    std::ptr::copy_nonoverlapping(src, dst, first_part);
+                } else {
+                    for i in 0..first_part {
+                        std::ptr::write(dst.add(i), T::default());
+                    }
+                }
```

Comment:
I feel like we can extract some of this repeated logic into a function for readability? (Rust should compile it the same.)